File: js/test/verdict-rules.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";

import {
  AndRule,
  FunctionRule,
  OrRule,
  RulesEngine,
} from "../dist/index.js";

/**
 * A rule that records every evaluation, so short-circuiting can be proven by
 * what actually ran rather than by the final boolean alone. A port that
 * evaluated concurrently would return the same boolean and fail only here.
 */
const counting = (name, passes, log, group) =>
  new FunctionRule(
    name,
    async () => {
      log.push(name);
      return { ruleName: name, passed: passes };
    },
    group,
  );

describe("structural typing", () => {
  it("a plain object of the right shape is a Rule, with nothing declared", async () => {
    // The property Python's Protocol gives and Dart/C# cannot: no class, no
    // `implements`, no registration — shape alone is enough.
    const duck = {
      name: "duck",
      async evaluate() {
        return { ruleName: "duck", passed: true };
      },
    };
    const result = await new AndRule("composed", [duck]).evaluate({});
    assert.equal(result.passed, true);
  });
});

describe("FunctionRule", () => {
  it("returns whatever the predicate returns, unchanged", async () => {
    const rule = new FunctionRule("r", async () => ({
      ruleName: "r",
      passed: true,
      detail: "why",
      data: { k: 1 },
    }));
    const result = await rule.evaluate({});
    assert.equal(result.detail, "why");
    assert.deepEqual(result.data, { k: 1 });
  });
});

describe("AndRule", () => {
  it("passes when every sub-rule passes, evaluating all of them", async () => {
    const log = [];
    const result = await new AndRule("all", [
      counting("a", true, log),
      counting("b", true, log),
    ]).evaluate({});
    assert.equal(result.passed, true);
    assert.deepEqual(log, ["a", "b"]);
  });

  it("short-circuits: later sub-rules never run", async () => {
    const log = [];
    const result = await new AndRule("all", [
      counting("a", true, log),
      counting("b", false, log),
      counting("c", true, log),
    ]).evaluate({});
    assert.equal(result.passed, false);
    assert.deepEqual(log, ["a", "b"], "'c' must never have been evaluated");
  });

  it("data holds only what ran, never padded, never flattened", async () => {
    const log = [];
    const result = await new AndRule("outer", [
      new AndRule("inner", [counting("deep", false, log)]),
      counting("never", true, log),
    ]).evaluate({});
    assert.equal(result.data.length, 1, "only the failing sub-rule ran");
    assert.equal(result.data[0].ruleName, "inner");
    assert.equal(
      result.data[0].data[0].ruleName,
      "deep",
      "nesting is preserved, not flattened into the parent",
    );
  });

  it("empty passes vacuously", async () => {
    assert.equal((await new AndRule("none", []).evaluate({})).passed, true);
  });
});

describe("OrRule", () => {
  it("short-circuits on the first pass", async () => {
    const log = [];
    const result = await new OrRule("any", [
      counting("a", false, log),
      counting("b", true, log),
      counting("c", true, log),
    ]).evaluate({});
    assert.equal(result.passed, true);
    assert.deepEqual(log, ["a", "b"], "'c' must never have been evaluated");
  });

  it("empty fails vacuously — the opposite polarity to AndRule", async () => {
    assert.equal((await new OrRule("none", []).evaluate({})).passed, false);
  });
});

describe("RulesEngine run modes", () => {
  it("runAll never short-circuits", async () => {
    const log = [];
    const engine = new RulesEngine([
      counting("a", false, log),
      counting("b", false, log),
      counting("c", true, log),
    ]);
    const result = await engine.runAll({});
    assert.equal(result.passed, false);
    assert.equal(result.results.length, 3);
    assert.deepEqual(log, ["a", "b", "c"]);
  });

  it("runGroup evaluates only its own group", async () => {
    const log = [];
    const engine = new RulesEngine([
      counting("a", true, log, "g1"),
      counting("b", true, log, "g2"),
      counting("c", false, log, "g1"),
    ]);
    const result = await engine.runGroup("g1", {});
    assert.deepEqual(result.results.map((r) => r.ruleName), ["a", "c"]);
    assert.equal(result.passed, false);
  });

  it("runNamed looks one rule up", async () => {
    const engine = new RulesEngine([counting("a", true, [])]);
    assert.equal((await engine.runNamed("a", {})).ruleName, "a");
  });
});

describe("emptiness is not absence", () => {
  it("an unknown rule name throws", async () => {
    const engine = new RulesEngine([counting("a", true, [], "g1")]);
    await assert.rejects(() => engine.runNamed("nope", {}), /No rule named/);
  });

  it("an unknown group throws rather than passing vacuously", async () => {
    // A group exists only because some rule declared it, so a lookup that
    // matches nothing can only be a typo. Returning a pass would mean a
    // misspelled group silently approves.
    const engine = new RulesEngine([counting("a", true, [], "g1")]);
    await assert.rejects(
      () => engine.runGroup("no-such-group", {}),
      /No rules in group/,
    );
  });

  it("a rule with no group makes no group exist", async () => {
    const engine = new RulesEngine([counting("a", true, [])]);
    await assert.rejects(() => engine.runGroup("g1", {}), /No rules in group/);
  });

  it("but empty composites still fold to their identity", async () => {
    assert.equal((await new AndRule("none", []).evaluate({})).passed, true);
    assert.equal((await new OrRule("none", []).evaluate({})).passed, false);
  });
});

describe("introspection", () => {
  it("reports exactly what the lookups accept", async () => {
    const engine = new RulesEngine([
      counting("a", true, [], "g1"),
      counting("b", true, [], "g2"),
      counting("c", true, []),
    ]);
    assert.deepEqual(engine.ruleNames, ["a", "b", "c"]);
    assert.deepEqual(engine.groupNames, ["g1", "g2"]);
    for (const g of engine.groupNames) await engine.runGroup(g, {});
  });

  it("an empty engine reports nothing", () => {
    const engine = new RulesEngine([]);
    assert.deepEqual(engine.ruleNames, []);
    assert.deepEqual(engine.groupNames, []);
  });
});

File: dart/test/verdict_rules_test.dart
import 'package:test/test.dart';
import 'package:verdict_rules/verdict_rules.dart';

/// A rule that records every evaluation, so short-circuiting can be proven by
/// what actually ran rather than by the final boolean alone. A port that
/// evaluates concurrently returns the same boolean and fails only here.
FunctionRule counting(String name, bool passes, List<String> log,
        {String? group}) =>
    FunctionRule(
      name,
      (ctx) async {
        log.add(name);
        return RuleResult(ruleName: name, passed: passes);
      },
      group: group,
    );

/// A plain top-level function, declared as nothing in particular.
Future<RuleResult> hasQuorum(Map<String, Object?> ctx) async =>
    RuleResult(ruleName: 'quorum', passed: ctx.length >= 3);

void main() {
  group('FunctionRule', () {
    test('returns whatever the predicate returns, unchanged', () async {
      final rule = FunctionRule(
        'r',
        (ctx) async => RuleResult(
            ruleName: 'r', passed: true, detail: 'why', data: {'k': 1}),
      );
      final result = await rule.evaluate({});
      expect(result.passed, isTrue);
      expect(result.detail, 'why');
      expect(result.data, {'k': 1});
    });

    test('carries its group label', () {
      expect(FunctionRule('r', (ctx) async => RuleResult(ruleName: 'r', passed: true),
              group: 'g')
          .group, 'g');
    });
  });

  group('AndRule', () {
    test('passes when every sub-rule passes, and evaluates all of them',
        () async {
      final log = <String>[];
      final result = await AndRule('all', [
        counting('a', true, log),
        counting('b', true, log),
      ]).evaluate({});
      expect(result.passed, isTrue);
      expect(log, ['a', 'b']);
    });

    test('short-circuits: later sub-rules never run', () async {
      final log = <String>[];
      final result = await AndRule('all', [
        counting('a', true, log),
        counting('b', false, log),
        counting('c', true, log),
      ]).evaluate({});
      expect(result.passed, isFalse);
      expect(log, ['a', 'b'], reason: "'c' must never have been evaluated");
    });

    test('data holds only what ran, never padded, never flattened', () async {
      final log = <String>[];
      final result = await AndRule('outer', [
        AndRule('inner', [counting('deep', false, log)]),
        counting('never', true, log),
      ]).evaluate({});
      final top = result.data! as List<RuleResult>;
      expect(top, hasLength(1), reason: 'only the failing sub-rule ran');
      expect(top.single.ruleName, 'inner');
      final nested = top.single.data! as List<RuleResult>;
      expect(nested.single.ruleName, 'deep',
          reason: 'nesting is preserved, not flattened into the parent');
    });

    test('empty passes vacuously', () async {
      expect((await AndRule('none', []).evaluate({})).passed, isTrue);
    });
  });

  group('OrRule', () {
    test('short-circuits on the first pass', () async {
      final log = <String>[];
      final result = await OrRule('any', [
        counting('a', false, log),
        counting('b', true, log),
        counting('c', true, log),
      ]).evaluate({});
      expect(result.passed, isTrue);
      expect(log, ['a', 'b'], reason: "'c' must never have been evaluated");
    });

    test('empty fails vacuously — the opposite polarity to AndRule', () async {
      expect((await OrRule('none', []).evaluate({})).passed, isFalse);
    });
  });

  group('RulesEngine run modes', () {
    test('runAll never short-circuits', () async {
      final log = <String>[];
      final engine = RulesEngine([
        counting('a', false, log),
        counting('b', false, log),
        counting('c', true, log),
      ]);
      final result = await engine.runAll({});
      expect(result.passed, isFalse);
      expect(result.results, hasLength(3));
      expect(log, ['a', 'b', 'c'],
          reason: 'every rule reports even after an earlier failure');
    });

    test('runGroup evaluates only its own group, and never short-circuits',
        () async {
      final log = <String>[];
      final engine = RulesEngine([
        counting('a', true, log, group: 'g1'),
        counting('b', true, log, group: 'g2'),
        counting('c', false, log, group: 'g1'),
      ]);
      final result = await engine.runGroup('g1', {});
      expect(result.results.map((r) => r.ruleName), ['a', 'c']);
      expect(result.passed, isFalse);
    });

    test('runNamed looks one rule up', () async {
      final engine = RulesEngine([counting('a', true, [])]);
      expect((await engine.runNamed('a', {})).ruleName, 'a');
    });
  });

  group('Emptiness is not absence', () {
    test('an unknown rule name throws', () async {
      final engine = RulesEngine([counting('a', true, [], group: 'g1')]);
      expect(() => engine.runNamed('nope', {}), throwsArgumentError);
    });

    test('an unknown group throws rather than passing vacuously', () async {
      // A group exists only because some rule declared it, so a lookup that
      // matches nothing can only be a typo or a stale name. Returning a pass
      // would mean a misspelled group silently approves.
      final engine = RulesEngine([counting('a', true, [], group: 'g1')]);
      expect(() => engine.runGroup('no-such-group', {}), throwsArgumentError);
    });

    test('a rule with no group makes no group exist', () async {
      final engine = RulesEngine([counting('a', true, [])]);
      expect(() => engine.runGroup('g1', {}), throwsArgumentError);
    });

    test('but empty composites still fold to their identity', () async {
      expect((await AndRule('none', []).evaluate({})).passed, isTrue);
      expect((await OrRule('none', []).evaluate({})).passed, isFalse);
    });
  });

  group('Try lookups', () {
    // These exist because the engine cannot know what an absent group means.
    // For one consumer it is "no constraint applies, pass"; for another "skip
    // and do not count it"; for a third "the configuration is wrong, fail".
    // A library default would be right for one and wrong for the rest.

    test('returns the result when present', () async {
      final log = <String>[];
      final engine = RulesEngine([
        counting('a', true, log, group: 'g1'),
        counting('b', false, log, group: 'g1'),
      ]);

      final group = await engine.tryRunGroup('g1', {});
      expect(group, isNotNull);
      expect(group!.results.map((r) => r.ruleName), ['a', 'b']);
      expect(group.passed, isFalse);

      expect((await engine.tryRunNamed('a', {}))?.ruleName, 'a');
    });

    test('returns null when absent', () async {
      final engine = RulesEngine([counting('a', true, [], group: 'g1')]);
      expect(await engine.tryRunGroup('no-such-group', {}), isNull);
      expect(await engine.tryRunNamed('nope', {}), isNull);
    });

    test('null means absent, never failed', () async {
      // Collapsing the two would make a typo indistinguishable from a
      // legitimate rejection.
      final engine = RulesEngine([counting('present', false, [])]);
      expect((await engine.tryRunNamed('present', {}))?.passed, isFalse);
      expect(await engine.tryRunNamed('absent', {}), isNull);
    });

    test('the strict forms are the try forms plus an assertion', () async {
      // Asserting the relationship keeps the two from drifting: one lookup
      // path, and the strict form adds only the throw.
      final engine = RulesEngine([counting('a', true, [], group: 'g1')]);
      final strict = await engine.runGroup('g1', {});
      final lenient = await engine.tryRunGroup('g1', {});
      expect(lenient, isNotNull);
      expect(strict.passed, lenient!.passed);
      expect(strict.results.length, lenient.results.length);
    });

    test('the caller chooses the fallback', () async {
      final engine = RulesEngine([counting('a', true, [], group: 'present')]);
      final result = await engine.tryRunGroup('absent', {});

      expect(result?.passed ?? true, isTrue); // absence means no constraint
      expect(result?.passed ?? false, isFalse); // absence means misconfigured
      expect([result].whereType<RunResult>(), isEmpty); // skip it
      expect(() => engine.runGroup('absent', {}), throwsArgumentError);
    });
  });

  group('Introspection', () {
    test('reports exactly what the lookups accept', () async {
      final engine = RulesEngine([
        counting('a', true, [], group: 'g1'),
        counting('b', true, [], group: 'g2'),
        counting('c', true, []),
      ]);
      expect(engine.ruleNames, ['a', 'b', 'c']);
      expect(engine.groupNames, ['g1', 'g2']);
      for (final g in engine.groupNames) {
        await engine.runGroup(g, {});
      }
    });

    test('an empty engine reports nothing', () {
      final engine = RulesEngine([]);
      expect(engine.ruleNames, isEmpty);
      expect(engine.groupNames, isEmpty);
    });
  });

  test('a top-level function is a rule with nothing declared', () async {
    // Dart's structural typing for function types: hasQuorum was never
    // declared to be anything rule-shaped, and matching the signature is
    // enough. Passed as a tear-off, not a closure.
    final result = await AndRule('composed', [
      FunctionRule('quorum', hasQuorum),
    ]).evaluate({'a': 1, 'b': 2, 'c': 3});
    expect(result.passed, isTrue);
  });

  test('a plain class satisfying the contract works without subclassing',
      () async {
    final result = await AndRule('composed', [_Custom()]).evaluate({});
    expect(result.passed, isTrue);
  });
}

/// A rule shape owning its own name and group must say `implements Rule`,
/// because Dart has no structural typing for multi-member interfaces. Function
/// types are a different story — see the tear-off test above, where a plain
/// top-level function is a rule with nothing declared.
class _Custom implements Rule {
  @override
  String get name => 'custom';

  @override
  String? get group => null;

  @override
  Future<RuleResult> evaluate(Map<String, Object?> context) async =>
      RuleResult(ruleName: name, passed: true);
}
